source/CIS/cis-4.rst

```diff
@@ -0,0 +1,705 @@
+- The credential status MUST NOT be ``Expired`` if the field ``valid_until`` was not present in :ref:`CIS-4-CredentialInfo` when registering the credential.
+- The credential status MUST be ``Acive`` if the credential is not revoked, and does not qualify as ``Expired`` or ``NotActivated``.
+
+.. _CIS-4-functions-issuer:
+
+``issuer``
+^^^^^^^^^^
+
+Query the issuer's account address.
+
+Response
+~~~~~~~~
+
+The function output is the issuer's account address.
+It is serialized as :ref:`CIS-4-AccountAddress`.
+
+.. _CIS-4-functions-issuerMetadata:
+
+``issuerMetadata``
+^^^^^^^^^^^^^^^^^^
+
+Query the current token metadata URLs for a list of token IDs.
+
+Response
+~~~~~~~~
+
+The function output is the issuer's metadata URL.
+
+It is serialized as :ref:`CIS-2-MetadataUrl`.
+
+.. _CIS-4-functions-registerCredential:
+
+``registerCredential``
+^^^^^^^^^^^^^^^^^^^^^^
+
+Register a credential with the given ID.
+
+Parameter
+~~~~~~~~~
+
+The parameter is the credential ID and credential information that is used to create an entry in the registry.
+
+It is serialized as :ref:`CIS-4-CredentialID` (``credential_id``) followed by :ref:`CIS-4-CredentialInfo` (``credential_info``)::
+
+  RegisterCredentialParameter ::= (credential_id: CredentialID) (credential_info: CredentialInfo)
+
+Requirements
+~~~~~~~~~~~~
+
+- The credential registration request MUST fail if the credential ID is already present in the registry.
+- The credential status after registration MUST NOT be ``Revoked`` (See the possible values for the status in :ref:`CIS-4-CredentialStatus`).
+
+.. _CIS-4-functions-revokeCredentialIssuer:
+
+``revokeCredentialIssuer``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Revoke a credential by the issuer's request.
+The issuer is authorized to revoke a credential if the transaction sender's address is the same as the return value of :ref:`CIS-4-functions-issuer`.
+
+Parameter
+~~~~~~~~~
+
+The parameter is the credential ID :ref:`CIS-4-CredentialID` and optional string indicating the revocation reason.
+
+It is serialized as :ref:`CIS-4-CredentialID` followed by 1 byte to indicate whether a reason is included, if its value is 0, then no reason string present, if the value is 1 then the bytes corresponding to the reason string follow::
+
+  OptionReason ::= (0: Byte)
+                 | (1: Byte) (n: Byte) (reason_string: Byteⁿ)
+  RevokeCredentialIssuerParam ::= (credential_id: CredentialID) (reason: OptionReason)
+
+.. TODO: what kind of characters are allowed? ASCII, Unicode?
+
+Requirements
+~~~~~~~~~~~~
+
+- If revoked successfully, the credential status MUST change to ``Revoked`` (see :ref:`CIS-4-functions-credentialStatus`).
+- The revocation MUST fail if:
+    - The the transaction ``sender`` is not the issuer.
+    - The credential ID is not present in the registry.
```

Comment:
I've considered this in the beginning. After discussing it with @abizjak, it looks like requiring certain rejection codes restricts implementation. One needs to specify the exact situations when the codes are returned in order for this to be reliably used in the wallet.
After that, I've decided to not specify any rejection codes.

Reply:
This is based on the experience of using CIS2 contracts.

The codes are not used because they are not reliable in general. So it's hard to make use of them in a meaningful way, e.g., in a client that interacts with general CIS2 contracts.

Reply:
The issue being that the codes are only defined in the sense of "if code is X then this must have happened", but we do not prescribe an `if and only if` since it prescribes implementation too much. For example our contracts often fail with "serialization failure".

Reply:
I agree, that we don't want to add requirements for the error codes, but standardizing an error code and the meaning of it, is still useful. If you want your contract to give a good user experience, you would want to provide the right error information back to the user and having some standardized error codes for this helps.

Reply:
What I tried to write above is that I don't think it actually helps.

But if you want to do it I'm not against it, but for me there isn't much value in this rpecisely because you cannot rely on it.

Reply:
An application should not rely their logic on this, no. But in the case of a UI it is useful, the wallet could show a nice error message and the contract developer would be motivated to take advantage of this.
I guess the schema for error codes, sort of solves this in the general case.

Reply:
This is just a consideration

Reply:
It's hard for me to see which codes to standardize and which not. Issuance is usually issuer-specific, not much to standardize there; verifiers and holders (wallets) mostly fetch data, so, the only relevant error is `CredentialNotFound`.
Holders can revoke in the wallet, but it can fail for multiple reasons, and it's hard to fix the granularity in advance.

This essentially leaves us with  `CredentialNotFound`, which I can include in the standard. It would require a slight change in the credential registry example:  the order of enum constructors [here](https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/credential-registry/src/lib.rs#L172) to make the CIS-4 error have the first tag.
And, of course, we should provide a library in the future similar to CIS-2 that will help avoid error code clashes.

@abizjak @limemloh How does that sound?

Reply:
> the order of enum constructors [here](https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/credential-registry/src/lib.rs#L172) to make the CIS-4 error have the first tag.

And I'll probably need to do something similar to the CIS-2 library https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/concordium-cis2/src/lib.rs#L923

Reply:
I experimented with adding a specific rejection code to the implementation and it looks like it brings more issues than solves. One should make sure that for the `CredentialNotFound` case a specific rejection code is produced and all custom errors don't clash with it. Since there is no clear benefit from doing this, I'll drop this idea and settle on not including any rejection code requirements.